dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/message/ProgramMessageRequestParamsMapper.java

```diff
@@ -0,0 +1,109 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.controller.message;
+
+import static org.hisp.dhis.webapi.controller.tracker.export.RequestParamsValidator.validateDeprecatedParameter;
+
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import org.hisp.dhis.common.IdentifiableObjectManager;
+import org.hisp.dhis.common.IllegalQueryException;
+import org.hisp.dhis.common.UID;
+import org.hisp.dhis.feedback.BadRequestException;
+import org.hisp.dhis.feedback.ConflictException;
+import org.hisp.dhis.program.Enrollment;
+import org.hisp.dhis.program.Event;
+import org.hisp.dhis.program.message.ProgramMessageQueryParams;
+import org.hisp.dhis.util.ObjectUtils;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Zubair Asghar
+ */
+@Component
+@RequiredArgsConstructor
+public class ProgramMessageRequestParamsMapper {
+  private final IdentifiableObjectManager manager;
+
+  @Transactional(readOnly = true)
+  public ProgramMessageQueryParams ValidateAndMap(ProgramMessageRequestParams params)
+      throws ConflictException, BadRequestException {
+    if (params == null) {
+      throw new BadRequestException("ProgramMessageRequestParams cannot be null.");
+    }
+
+    UID enrollmentUid =
+        validateDeprecatedParameter(
+            "programInstance", params.getProgramInstance(), "enrollment", params.getEnrollment());
+    UID eventUid =
+        validateDeprecatedParameter(
+            "programStageInstance", params.getProgramStageInstance(), "event", params.getEvent());
+
+    if (enrollmentUid == null && eventUid == null) {
+      throw new ConflictException("Enrollment or Event must be specified.");
+    }
+    if (enrollmentUid != null && eventUid != null) {
+      throw new ConflictException("Enrollment and Event cannot be processed together.");
+    }
+
+    UID entity = ObjectUtils.firstNonNull(enrollmentUid, eventUid);
+
+    Enrollment enrollment = null;
+    Event event = null;
+
+    if (enrollmentUid != null) {
+      enrollment =
+          Optional.ofNullable(manager.get(Enrollment.class, entity.getValue()))
+              .orElseThrow(
+                  () ->
+                      new IllegalQueryException(
```

Comment:
Would it make sense to be `NotFoundException` instead?

Reply:
absolutely but it will change the API response.